### ISSUE TITLE
install tensorflow 2.3.0 gpu version

### ISSUE THREAD

**Issue by zouw2** [rstudio/tensorflow#445]
Hi, I am trying to install tensorflow 2.3.0 with gpu support
library(tensorflow)
install_tensorflow(version = "2.3.0-gpu")

but it seems that the tensorflow version 2.2.0 is eventually being installed.

I need tensorflow 2.3.0 because the current r package tfprobability requires tensorflow version 2.3.0 
thanks
wei

**Comment by zouw2** [comment]
I am realizing that it might be wrong to figure out the version of tensorflow using the R command installed.packages()['tensorflow', ]. What's the correct way here? thanks
wei

**Comment by anirban-mukherjee** [comment]
Is there a way we can find out which TF releases (min release, max release) are supported by the tensorflow R package? For example, I am presuming TF 2.3 and TF 2.4 are not supported as the package installs TF 2.2. Is that correct?

A suggestion: it might be useful if there was an issues tracker for each TF release where we could file (beta test) bug reports. My interest in TF 2.4, for example, stems from Apple's optimised TF fork (for Mac) being on TF 2.4. I do want to try it out but I don't want to do so if there are known problems. A mean (mailing list, forum) to exchange such information would be very useful!

**Comment by anirban-mukherjee** [comment]
@zouw2 You can use your python package manager to update to TF 2.3.x (and more generally to update TF). For example, you can use pip to update to TF 2.3.1. The R package connects to the Python package using Reticulate. Therefore, the version of TF is in your Python installation (or virtual environment if you are using a virtual environment) is the version R will see and use. Please be aware though that I cannot tell if the R package has been tested against the more recent versions of TF. TF tends to make breaking changes so I would not do so for a production machine but perhaps thats ok for a test machine or something one is learning on.

**Comment by zouw2** [comment]
thanks! I noted that rstudio's recent [AI blogs](https://blogs.rstudio.com/ai/) are all about torch. Does rstudio plan to mostly focus on torch in the future?
wei

**Comment by touala** [comment]
Hi,

Could someone with more expertise than me confirm what @anirban-mukherjee is saying about potential incompatibilities between the released `R` package (`2.2.0`) and the nightly build `python` package (`2.5.0`)? I would like to use CUDA11 features but I cannot get it to work from R.

Thanks,

Alan

**Comment by anirban-mukherjee** [comment]
2.2.x are the older releases, 2.3.1 is the current release, and 2.4.0 is in RC. 2.5.0 is the unstable (dev) version that comes with no guarantee of stability (as of today). I would not suggest anything higher than the release version (2.3.1) unless you absolutely need to. In particular, the nightly is guaranteed to give you all kinds of trouble. 

In my case, I want to use Apple's new binary, which has only been released for 2.4.0. Apple has not released 2.2.x and 2.3.x binaries. FWIW, I have been using the 2.4.0 binary and (mostly) so far so good (and it is significantly faster than a binary compiled from source on my system). I think _there_ is a bug with the verbose option in keras::fit (when using RStudio on Mac) but I cannot narrow it down to a MWE. Instead, I set verbose to 0 and so far so good.

You should be able to do most things with 2.2.0 and not need 2.3.1 so unless you need 2.3.1 for TensorFlow probability or another library, first try using 2.2.0.

The R package uses the TF binary in Python through a library called reticulate. The R tensorflow package provides syntactic sugar that makes it easier to work with TensorFlow for a R programmer. TF makes significant (breaking) changes across API releases. Therefore, it is possible (and even likely) that some functions in the R tensorflow package will not work across all TF API releases. For e.g., [issue with setting seed](https://github.com/rstudio/tensorflow/issues/445). Furthermore, some of the tutorials and examples may also not work in newer TF releases if the tutorial was written using an older API. In those cases, you can use always use tf to access the Python TensorFlow module in Python and directly use the newer TF API. Ultimately, reticulate will work regardless of the R tensorflow package or the TF release.

HTH.

**Comment by touala** [comment]
Thank you very much @anirban-mukherjee for those advices.

Alan

**Comment by zouw2** [comment]
thank you,  @anirban-mukherjee !
wei

**Comment by t-kalinowski** [comment]
Hello,

The code that installs tensorflow and keras has been [refactored](https://github.com/rstudio/tensorflow/pull/476).
Please update to the latest versions of reticulate, tensorflow, and keras:
```R
install.packages("remotes")
remotes::install_github(paste0("rstudio/", c("reticulate", "tensorflow", "keras")))
reticulate::install_miniconda() # skip this if you want to self-install conda or use venv
keras::install_keras()
```

Please let us know if after updating you are still encountering installation issues:
https://github.com/rstudio/tensorflow/issues/new
